File: Projects/Testtrack/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/infrastructure/promotion/CheapestProductPromotion.java
package com.simple.ecommerce.infrastructure.promotion;

import com.simple.ecommerce.application.port.PromotionStrategy;
import com.simple.ecommerce.domain.entity.Product;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

/**
 * A promotion that applies a special price to every nth product in the cart.
 * Business Rules:
 * - Every third product costs 1 PLN
 * - Applied to cheapest products first
 */
public class CheapestProductPromotion implements PromotionStrategy {
    private static final int DEFAULT_PRODUCTS_REQUIRED = 3;
    private static final double DEFAULT_SPECIAL_PRICE = 1.0;

    private final String promotionCode;
    private final double specialPrice;
    private final int productsRequired;

    /**
     * Creates a promotion where every third product costs 1 PLN.
     *
     * @param promotionCode the promotion code
     */
    public CheapestProductPromotion(String promotionCode) {
        this(promotionCode, DEFAULT_SPECIAL_PRICE, DEFAULT_PRODUCTS_REQUIRED);
    }

    /**
     * Creates a promotion with custom parameters.
     *
     * @param promotionCode the promotion code
     * @param specialPrice the special price for the cheapest product
     * @param productsRequired the number of products required to activate the promotion
     */
    public CheapestProductPromotion(String promotionCode, double specialPrice, int productsRequired) {
        if (promotionCode == null || promotionCode.trim().isEmpty()) {
            throw new IllegalArgumentException("Promotion code cannot be empty");
        }
        if (specialPrice < 0) {
            throw new IllegalArgumentException("Special price cannot be negative");
        }
        if (productsRequired < 2) {
            throw new IllegalArgumentException("Products required must be at least 2");
        }

        this.promotionCode = promotionCode;
        this.specialPrice = specialPrice;
        this.productsRequired = productsRequired;
    }

    @Override
    public double calculateDiscount(Map<Product, Integer> products) {
        if (products.isEmpty()) {
            return 0;
        }

        // Expand the cart into a list of individual products
        List<Product> allProducts = new ArrayList<>();
        products.forEach((product, quantity) -> {
            for (int i = 0; i < quantity; i++) {
                allProducts.add(product);
            }
        });

        // If we don't have enough products, no discount applies
        if (allProducts.size() < productsRequired) {
            return 0;
        }

        // Sort products by price (cheapest first)
        allProducts.sort(Comparator.comparing(Product::getPrice));

        double totalDiscount = 0;

        // Apply the discount to every Nth product (where N = productsRequired)
        for (int i = 0; i < allProducts.size(); i++) {
            if ((i + 1) % productsRequired == 0) { // Every Nth product gets the discount
                Product discountedProduct = allProducts.get(i);
                // Discount is the difference between original price and special price
                double productDiscount = discountedProduct.getPrice() - specialPrice;
                // Ensure discount is not negative (special price doesn't exceed product price)
                if (productDiscount > 0) {
                    totalDiscount += productDiscount;
                }
            }
        }

        return totalDiscount;
    }

    @Override
    public String getPromotionCode() {
        return promotionCode;
    }
}

File: Projects/Testtrack/p911 - Simple eCommerce backend/infrastructure/promotion/cheapest_product_promotion.py
from typing import Dict, List
from application import PromotionStrategy
from domain import Product


class CheapestProductPromotion(PromotionStrategy):
    DEFAULT_PRODUCTS_REQUIRED = 3
    DEFAULT_SPECIAL_PRICE = 1.0

    def __init__(self, promotion_code: str, special_price: float = None, products_required: int = None):
        if special_price is None:
            special_price = self.DEFAULT_SPECIAL_PRICE
        if products_required is None:
            products_required = self.DEFAULT_PRODUCTS_REQUIRED

        if not promotion_code or promotion_code.strip() == "":
            raise ValueError("Promotion code cannot be empty")
        else:
            self.promotion_code = promotion_code

        if special_price < 0:
            raise ValueError("Special price cannot be negative")
        else:
            self.special_price = special_price

        if products_required < 2:
            raise ValueError("Products required must be at least 2")
        else:
            self.products_required = products_required

    def calculate_discount(self, products: Dict[Product, int]) -> float:
        if not products:
            return 0

        all_products: List[Product] = []
        for product, quantity in products.items():
            for _ in range(quantity):
                all_products.append(product)

        if len(all_products) < self.products_required:
            return 0

        all_products.sort(key=lambda p: p.get_price())

        total_discount = 0
        for i in range(len(all_products)):
            if (i + 1) % self.products_required == 0:
                discounted_product = all_products[i]
                product_discount = discounted_product.get_price() - self.special_price
                if product_discount > 0:
                    total_discount += product_discount

        return total_discount

    def get_promotion_code(self) -> str:
        return self.promotion_code
